python/src/iceberg/expressions/base.py

```diff
@@ -122,3 +125,143 @@ def __le__(self, other):
 
     def __ge__(self, other):
         return self.value >= other.value
+
+
+class BooleanExpression(ABC):
+    """base class for all boolean expressions"""
+
+    @abstractmethod
+    def __invert__(self) -> "BooleanExpression":
+        ...
+
+
+class And(BooleanExpression):
+    """AND operation expression - logical conjunction"""
+
+    def __new__(cls, left: BooleanExpression, right: BooleanExpression, *rest: BooleanExpression):
+        if rest:
+            return reduce(And, (left, right, *rest))
+        if left is AlwaysFalse() or right is AlwaysFalse():
+            return AlwaysFalse()
+        elif left is AlwaysTrue():
+            return right
+        elif right is AlwaysTrue():
+            return left
+        return super().__new__(cls)
+
+    def __init__(
+        self,
+        left: BooleanExpression,
+        right: BooleanExpression,
+        *rest: BooleanExpression,
+    ):
+        if not rest:
```

Comment:
Looks like the tests are failing on the new `ValueError`. Maybe updating this to `if not rest or len(rest) <= 0:`?